jax/_src/scipy/stats/_core.py

```diff
@@ -0,0 +1,86 @@
+# Copyright 2022 The JAX Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import namedtuple
+from functools import partial
+from typing import Optional, Tuple
+
+import jax.numpy as jnp
+import scipy
+from jax import jit
+from jax._src.api import vmap
+from jax._src.numpy.lax_numpy import _check_arraylike
+from jax._src.numpy.util import _wraps
+from jax._src.typing import ArrayLike
+from jax._src.util import canonicalize_axis
+
+ModeResult = namedtuple('ModeResult', ('mode', 'count'))
+
+@_wraps(scipy.stats.mode, lax_description="""\
+Currently the only supported nan_policy is 'propagate'
+""")
+@partial(jit, static_argnames=['axis', 'nan_policy', 'keepdims'])
+def mode(a: ArrayLike, axis: Optional[int] = 0, nan_policy: str = "propagate", keepdims: bool = False) -> ModeResult:
+  _check_arraylike("mode", a)
+  x = jnp.atleast_1d(a)
+
+  if nan_policy not in ["propagate", "omit", "raise"]:
+    raise ValueError(
+      f"Illegal nan_policy value {nan_policy!r}; expected one of "
+      "{'propoagate', 'omit', 'raise'}"
+    )
+  if nan_policy == "omit":
+    # TODO: return answer without nans included.
+    raise NotImplementedError(
+      f"Logic for `nan_policy` of {nan_policy} is not implemented"
+    )
+  if nan_policy == "raise":
+    raise NotImplementedError(
+      "In order to best JIT compile `mode`, we cannot know whether `x` contains nans. "
+      "Please check if nans exist in `x` outside of the `mode` function."
+    )
+
+  if x.size == 0:
+    if axis is None:
+      if keepdims:
+        output_shape = [1 for _ in range(x.ndim)]
+        return ModeResult(jnp.full(output_shape, jnp.nan), jnp.full(output_shape, jnp.nan))
+      else:
+        return ModeResult(jnp.nan, jnp.nan)
+    else:
+      output_shape = list(x.shape)
+      if keepdims:
+        output_shape[axis] = 1
+      else:
+        del output_shape[axis]
+      return ModeResult(jnp.full(output_shape, jnp.nan), jnp.full(output_shape, jnp.nan))
+  if axis is None:
+    x = jnp.ravel(x)
+    axis = 0
+
+  def _mode_helper(x: jnp.ndarray) -> Tuple[jnp.ndarray, jnp.ndarray]:
+    """Helper function to return mode and count of a given array."""
+    vals, counts = jnp.unique(x, return_counts=True, size=x.size)
+    return vals[jnp.argmax(counts)], counts.max()
+
+  axis = canonicalize_axis(axis, x.ndim)
+  output_shape = list(x.shape)
+  if keepdims:
+    output_shape[axis] = 1
+  else:
+    del output_shape[axis]
```

Comment:
Does this result in the correct shape for `axis=None` with `keepdims=True`?

Reply:
Yes, I think this part of the code is fine for axis=None with keepdims=True.